UrkelTree.Commit() crashes, if no new keys/values inserted

If existing keys with existing values are re-inserted to Urkel, Commit() crashes with the following message:
```
panic: interface conversion: node.Node is nil, not *node.LeafNode [recovered]
        panic: interface conversion: node.Node is nil, not *node.LeafNode

goroutine 19 [running]:
testing.tRunner.func1(0xc000170400)
        /opt/go/src/testing/testing.go:830 +0x392
panic(0xa11340, 0xc0002d6300)
        /opt/go/src/runtime/panic.go:522 +0x1b5
github.com/oasislabs/ekiden/go/storage/mkvs/urkel/db/memory.(*memoryBatch).PutWriteLog(0xc0002d07c0, 0x82b5f5d50d09e171, 0xb807690b50906041, 0x387e9f0eca63e795, 0x1bba3c032474e8e4, 0x0, 0xb8ab0f3dfa1692f3, 0x1d9bb3302c49fb9b, 0x296e8908b98155d8, 0x5bdf272f50c08c1, ...)
        /home/oa/ekiden/go/storage/mkvs/urkel/db/memory/memory.go:173 +0x56d
github.com/oasislabs/ekiden/go/storage/mkvs/urkel.(*Tree).commitWithHooks(0xc000122e80, 0xba5c80, 0xc0000280d8, 0x82b5f5d50d09e171, 0xb807690b50906041, 0x387e9f0eca63e795, 0x1bba3c032474e8e4, 0x0, 0x0, 0x0, ...)
        /home/oa/ekiden/go/storage/mkvs/urkel/urkel.go:354 +0x8d1
github.com/oasislabs/ekiden/go/storage/mkvs/urkel.(*Tree).Commit(...)
        /home/oa/ekiden/go/storage/mkvs/urkel/urkel.go:298
github.com/oasislabs/ekiden/go/storage/mkvs/urkel.testNodeEviction(0xc000170400, 0xbab440, 0xc00011c4c0)
        /home/oa/ekiden/go/storage/mkvs/urkel/urkel_test.go:519 +0x5a6
github.com/oasislabs/ekiden/go/storage/mkvs/urkel.testBackend.func11(0xc000170400)
        /home/oa/ekiden/go/storage/mkvs/urkel/urkel_test.go:671 +0xb4
testing.tRunner(0xc000170400, 0xc000151400)
        /opt/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
        /opt/go/src/testing/testing.go:916 +0x35a
exit status 2
```

## Acceptance Criteria

Commit() should not cause panic ;)

## Testing

To test the fix, change `GenerateKeyValuePairsEx` so it doesn't take the `prefix` argument:
```
diff --git a/go/storage/mkvs/urkel/urkel_test.go b/go/storage/mkvs/urkel/urkel_test.go
index 57537deb..2ce283b5 100644
--- a/go/storage/mkvs/urkel/urkel_test.go
+++ b/go/storage/mkvs/urkel/urkel_test.go
@@ -894,8 +894,8 @@ func generateKeyValuePairsEx(prefix string, count int) ([][]byte, [][]byte) {
        keys := make([][]byte, count)
        values := make([][]byte, count)
        for i := 0; i < count; i++ {
-               keys[i] = []byte(fmt.Sprintf("%skey %d", prefix, i))
-               values[i] = []byte(fmt.Sprintf("%svalue %d", prefix, i))
+               keys[i] = []byte(fmt.Sprintf("%skey %d", "", i))
+               values[i] = []byte(fmt.Sprintf("%svalue %d", "", i))
        }
 
        return keys, values
```

This will cause insertion of a bunch of same keys/vals. If the fix works, `testNodeEviction` should not crash anymore (but only complain about different internal nodes, because of the same keys).

--- a/go/storage/mkvs/urkel/db/memory/memory.go
+++ b/go/storage/mkvs/urkel/db/memory/memory.go
@@ -168,9 +168,21 @@ func (b *memoryBatch) PutWriteLog(
 	digest := make(writeLogDigest, len(writeLog))
 	for idx, entry := range writeLog {
 		if annotations[idx].InsertedNode != nil {
+			nd := annotations[idx].InsertedNode.Node
+			if nd == nil {
+				raw, err := b.db.getLocked(annotations[idx].InsertedNode.Hash)
+				if err != nil {
+					return err
+				}
+
+				nd, err = node.UnmarshalBinary(raw)
+				if err != nil {
+					return err
+				}
+			}
 			digest[idx] = logEntryDigest{
 				key:  entry.Key,
-				leaf: annotations[idx].InsertedNode.Node.(*node.LeafNode),
+				leaf: nd.(*node.LeafNode),
 			}
 		} else {
 			digest[idx] = logEntryDigest{

--- a/go/storage/mkvs/urkel/urkel_test.go
+++ b/go/storage/mkvs/urkel/urkel_test.go
@@ -527,6 +527,29 @@ func testNodeEviction(t *testing.T, ndb db.NodeDB) {
 	require.EqualValues(t, 676, stats.Cache.LeafValueSize, "Cache.LeafValueSize")
 }
 
+func testDoubleInsertWithEviction(t *testing.T, ndb db.NodeDB) {
+	ctx := context.Background()
+	tree := New(nil, ndb, Capacity(128, 0))
+
+	keys, values := generateKeyValuePairs()
+	for i := 0; i < len(keys); i++ {
+		err := tree.Insert(ctx, keys[i], values[i])
+		require.NoError(t, err, "Insert")
+	}
+
+	_, _, err := tree.Commit(ctx, testNs, 0)
+	require.NoError(t, err, "Commit")
+
+	keys, values = generateKeyValuePairs()
+	for i := 0; i < len(keys); i++ {
+		err = tree.Insert(ctx, keys[i], values[i])
+		require.NoError(t, err, "Insert")
+	}
+
+	_, _, err = tree.Commit(ctx, testNs, 0)
+	require.NoError(t, err, "Commit")
+}
+
 func testDebugDump(t *testing.T, ndb db.NodeDB) {
 	ctx := context.Background()
 	tree := New(nil, ndb)
@@ -669,6 +692,11 @@ func testBackend(t *testing.T, initBackend func(t *testing.T) (db.NodeDB, interf
 		backend, custom := initBackend(t)
 		defer finiBackend(t, backend, custom)
 		testNodeEviction(t, backend)
+	})
+	t.Run("DoubleInsertWithEviction", func(t *testing.T) {
+		backend, custom := initBackend(t)
+		defer finiBackend(t, backend, custom)
+		testDoubleInsertWithEviction(t, backend)
 	})
 	t.Run("DebugDump", func(t *testing.T) {
 		backend, custom := initBackend(t)

